csharp/src/Apache.Arrow/Field.Builder.cs

```diff
@@ -52,6 +55,131 @@ public Builder DataType(IArrowType type)
+                        break;
+                    case var _ when valueType == typeof(ulong):
+                        DataType(new UInt64Type());
+                        break;
+                    // Decimal
+                    case var _ when valueType == typeof(decimal):
+                        DataType(Decimal256Type.Default);
+                        break;
+                    // Float
+                    case var _ when valueType == typeof(float):
+                        DataType(new FloatType());
+                        break;
+                    // Double
+                    case var _ when valueType == typeof(double):
+                        DataType(new DoubleType());
+                        break;
+                    // DateTime
+                    case var _ when valueType == typeof(DateTime) || valueType == typeof(DateTimeOffset):
+                        DataType(new TimestampType(TimeUnit.Nanosecond, timezone));
+                        break;
+                    // Time
+                    case var _ when valueType == typeof(TimeSpan):
+                        DataType(new Time64Type(TimeUnit.Nanosecond));
+                        break;
+#if NETCOREAPP3_1_OR_GREATER
+                    // Dictionary: IDictionary<?,?>
+                    case var dict when typeof(IDictionary).IsAssignableFrom(valueType):
+                        genericArgs = dict.GetGenericArguments();
+
+                        DataType(new DictionaryType(
+                            new Builder().DataType(genericArgs[0], timezone)._type,
+                            new Builder().DataType(genericArgs[1], timezone)._type,
+                            false
+                        ));
+
+                        Nullable(true);
+
+                        break;
+                    // IEnumerable: List, Array, ...
+                    case var list when typeof(IEnumerable).IsAssignableFrom(valueType):
+                        elementType = list.GetGenericArguments()[0];
+
+                        DataType(new ListType(new Builder().Name("item").DataType(elementType, timezone).Build()));
+
+                        Nullable(true);
+                        break;
+                    // Struct like: get all properties
+                    case var structure when (valueType.IsValueType && !valueType.IsEnum && !valueType.IsPrimitive):
+                        if (string.IsNullOrEmpty(_name))
+                            Name(structure.Name);
+
+                        PropertyInfo[] properties = structure.GetProperties(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic);
+
+                        DataType(new StructType(
+                            properties
+                                .Select(property => new Builder().Name(property.Name).DataType(property.PropertyType, timezone).Build())
+                                .ToArray()
+                        ));
+
+                        break;
```

Comment:
This seems rather powerful, though also a little dangerous :)

Reply:
Apache Spark uses to same process in scala, its pretty usefull when later you want statically type your Datasets which can be considered like arrow RecordBatches